src/Shared/ThrowHelpers/ArgumentOutOfRangeThrowHelper.cs

```diff
@@ -0,0 +1,168 @@
+    [DoesNotReturn]
+    private static void ThrowNegativeOrZero<T>(string? paramName, T value)
+    {
+        throw new ArgumentOutOfRangeException(paramName, value, $"'{paramName}' must be a non-negative and non-zero value.");
+    }
+
+    [DoesNotReturn]
+    private static void ThrowGreater<T>(string? paramName, T value, T other)
+    {
+        throw new ArgumentOutOfRangeException(paramName, value, $"'{paramName}' must be less than or equal to '{other}'.");
+    }
+
+    [DoesNotReturn]
+    private static void ThrowGreaterEqual<T>(string? paramName, T value, T other)
+    {
+        throw new ArgumentOutOfRangeException(paramName, value, $"'{value}' must be less than '{other}'.");
+    }
+
+    [DoesNotReturn]
+    private static void ThrowLess<T>(string? paramName, T value, T other)
+    {
+        throw new ArgumentOutOfRangeException(paramName, value, $"'{value}' must be greater than or equal to '{other}'.");
+    }
+
+    [DoesNotReturn]
+    private static void ThrowLessEqual<T>(string? paramName, T value, T other)
+    {
+        throw new ArgumentOutOfRangeException(paramName, value, $"'{value}' must be greater than '{other}'.");
+    }
+#endif
+
+    /// <summary>Throws an <see cref="ArgumentOutOfRangeException"/> if <paramref name="value"/> is zero.</summary>
+    /// <param name="value">The argument to validate as non-zero.</param>
+    /// <param name="paramName">The name of the parameter with which <paramref name="value"/> corresponds.</param>
+    public static void ThrowIfZero(int value, [CallerArgumentExpression(nameof(value))] string? paramName = null)
+    {
+#if !NET7_0_OR_GREATER
+        if (value == 0)
+        {
+            ThrowZero(paramName, value);
+        }
+#else
+        ArgumentOutOfRangeException.ThrowIfZero(value, paramName);
+#endif
+    }
+
+    /// <summary>Throws an <see cref="ArgumentOutOfRangeException"/> if <paramref name="value"/> is negative.</summary>
+    /// <param name="value">The argument to validate as non-negative.</param>
+    /// <param name="paramName">The name of the parameter with which <paramref name="value"/> corresponds.</param>
+    public static void ThrowIfNegative(int value, [CallerArgumentExpression(nameof(value))] string? paramName = null)
+    {
+#if !NET7_0_OR_GREATER
+        if (value < 0)
+        {
+            ThrowNegative(paramName, value);
+        }
+#else
+        ArgumentOutOfRangeException.ThrowIfNegative(value, paramName);
+#endif
+    }
+
+    /// <summary>Throws an <see cref="ArgumentOutOfRangeException"/> if <paramref name="value"/> is negative or zero.</summary>
+    /// <param name="value">The argument to validate as non-zero or non-negative.</param>
+    /// <param name="paramName">The name of the parameter with which <paramref name="value"/> corresponds.</param>
+    public static void ThrowIfNegativeOrZero(int value, [CallerArgumentExpression(nameof(value))] string? paramName = null)
+    {
+#if !NET7_0_OR_GREATER
+        if (value < 0 || value == 0)
```

Comment:
compiler doesn't seem to reduce this to `<=` - IL looks simpler with `<=` ? https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBLANgHwAEAmARgFgAoQgZgAIS6BhOgbyrs4fuAglzoBZABTYAdhjoBDAJScAvAD5pdADzy6ABgDcHLrTq9+dAHKiJ0uXSUrVWuvnwr5GnVQC+QA

Reply:
C# compiler doesn't. JIT will with https://github.com/dotnet/runtime/pull/78786. But if you can use a single comparison instead of two, yeah, better to do so.

Reply:
since this only applies to down-level runtimes, we should presume the JIT is not so savvy

Reply:
```suggestion
        if (value <= 0)
```